src/cfnlint/rules/resources/NoEcho.py

```diff
@@ -0,0 +1,51 @@
+"""
+Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+SPDX-License-Identifier: MIT-0
+"""
+from cfnlint.helpers import bool_compare
+from cfnlint.rules import CloudFormationLintRule
+from cfnlint.rules import RuleMatch
+
+
+class NoEcho(CloudFormationLintRule):
+    id = 'W4002'
+    shortdesc = 'Check for NoEcho References'
+    description = 'Check if there is a NoEcho enabled parameter referenced within a resources Metadata section'
+    source_url = 'https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/parameters-section-structure.html#parameters-section-structure-properties'
+    tags = ['resources', 'NoEcho']
+
+    def match(self, cfn):
+        matches = []
+        no_echo_params = []
+        parameters = cfn.get_parameters()
+        for parameter_name, parameter_value in parameters.items():
+            noecho = parameter_value.get('NoEcho', default=False)
+            if bool_compare(noecho, True):
+                no_echo_params.append(parameter_name)
+
+        if not no_echo_params:
+            return no_echo_params
+
+        resource_properties = cfn.get_resources()
+        resource_dict = {key: resource_properties[key] for key in resource_properties if
+                         isinstance(resource_properties[key], dict)}
+        for resource_name, resource_values in resource_dict.items():
+            resource_values = {key: resource_values[key] for key in resource_values if
+                               isinstance(resource_values[key], dict)}
+            for key, metadata in resource_values.items():
+                if key == 'Metadata':
+                    metadata = {key: metadata[key] for key in metadata if
+                                isinstance(metadata[key], dict)}
+                    for prop_name, properties in metadata.items():
+                        if isinstance(properties, dict):
```

Comment:
Is this check necessary if you are already doing `if isinstance(metadata[key], dict)` on line 37,38

Reply:
Would be a redundant check - Left as is since one of the checks was removed while implementing the above changes.